cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -313,21 +289,32 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
       }
       DatasetId datasetId = DatasetId.fromString(datasetIdString);
       for (ProgramSchedule schedule : getSchedules(context, Schedulers.triggerKeyForPartition(datasetId))) {
-        addJob(new Job(schedule));
+        JobQueueDataset jobQueue = getJobQueue(context);
+        boolean existingJobModified = false;
+        try (CloseableIterator<Job> jobs = jobQueue.getJobsForSchedule(schedule.getScheduleId())) {
+          while (jobs.hasNext()) {
+            Job job = jobs.next();
+            if (job.getState() != Job.State.PENDING_LAUNCH) {
+              // check if its PENDING_LAUNCH. If so, create a new Job to avoid the chance that the job-launching
+              // process has conflict
+              jobQueue.addNotification(job, notification);
+              existingJobModified = true;
+            }
+          }
+        }
+        // if no existing job was modified with the new notification, add a new job with the first notification
+        if (!existingJobModified) {
+          Job job = new SimpleJob(schedule, System.currentTimeMillis(),
+                                  Lists.newArrayList(notification), Job.State.PENDING_TRIGGER);
+          jobQueue.put(job);
+        }
       }
     }
   }
 
-  private static final class Job {
-    ProgramSchedule schedule;
-
-    Job(ProgramSchedule schedule) {
-      this.schedule = schedule;
-    }
-
-    public ProgramSchedule getSchedule() {
-      return schedule;
-    }
+  private JobQueueDataset getJobQueue(DatasetContext context) throws IOException, DatasetManagementException {
```

Comment:
why not save `JobQueueDataset` as a field and in `CoreSchedulerSerivce` wait for `JobQueueDataset` to be created before starting `NotificationSubscriberService` and `ConstraintCheckerService`?